desktop_version/src/LevelDebugger.cpp

```diff
@@ -0,0 +1,459 @@
+#include "LevelDebugger.h"
+
+#include "Constants.h"
+#include "Entity.h"
+#include "Font.h"
+#include "Graphics.h"
+#include "KeyPoll.h"
+#include "Localization.h"
+#include "Map.h"
+#include "Screen.h"
+#include "Script.h"
+#include "UtilityClass.h"
+#include "VFormat.h"
+
+namespace level_debugger
+{
+    bool active = false;
+    bool should_pause = true;
+    bool tab_held = false;
+    bool debug_held = false;
+    bool forced = false;
+
+    // Moving entities/blocks
+    bool mouse_held = false;
+    int held_entity = -1;
+    int held_block = -1;
+    int grabber_offset_x = 0;
+    int grabber_offset_y = 0;
+    int mx = 0;
+    int my = 0;
+
+    bool right_mouse_held = false;
+
+    bool is_pausing(void)
+    {
+        return active && should_pause;
+    }
+
+    bool is_active(void)
+    {
+        return active;
+    }
+
+    void toggle_active(void)
+    {
+        active = !active;
+    }
+
+    bool mouse_within(SDL_Rect* rect)
+    {
+        SDL_Point mouse = { mx, my };
+
+        return SDL_PointInRect(&mouse, rect);
+    }
+
+    void set_forced(void)
+    {
+        forced = true;
+    }
+
+    void input(void)
+    {
+        if (!forced && (!map.custommode || map.custommodeforreal))
+        {
+            active = false;
+            return;
+        }
+
+        if (key.isDown(SDLK_y))
+        {
+            if (!debug_held)
+            {
+                debug_held = true;
+                active = !active;
+            }
+        }
+        else
+        {
+            debug_held = false;
+        }
+
+        if (!active)
+        {
+            return;
+        }
+
+        if (key.isDown(SDLK_TAB))
+        {
+            if (!tab_held)
+            {
+                tab_held = true;
+                should_pause = !should_pause;
+            }
+        }
+        else
+        {
+            tab_held = false;
+        }
+
+        mx = key.mx;
+        my = key.my;
+
+        if (gameScreen.scalingMode == SCALING_STRETCH) {
+            // In this mode specifically, we have to fix the mouse coordinates
+            int screenwidth, screenheight;
+            gameScreen.GetScreenSize(&screenwidth, &screenheight);
+            mx = mx * 320 / screenwidth;
+            my = my * 240 / screenheight;
+        }
```

Comment:
I didn't want you to just copy-paste the code. So nevermind; I committed 8e3e29a14c98a8a4113d565c6c131c389eb3705d, which removes the need for this to be here entirely. Please rebase but without this scaling fix code.